Add support for windows per-host-subnet

orangedeng opened the pull request:
refer to https://github.com/rancher/rancher/issues/10447
Using powershell `*-netroute` to modify route rules in windows.
Using netsh tool to add nat port mapping rules.
Add support for windows service.

Commit:
Add CROSS compile and startup script for windows per-host-subnet

Add a script file for windows per-host-subnet and add it to package
script.
Change Dockerfile.dapper for adding zip support

Commit:
Add windows route support

Use powershell command to add/remove route for other hosts' subnet.

Commit:
Add vendors for windows per-host-subnet support

Adding new vendors and change vendor.conf

Commit:
Add port mapping support in windows per-host-subnet

Use netsh to add port mapping rules on nat adapters.

orangedeng commented:
@yasker

yasker commented:
@leodotcloud can you take a look?

leodotcloud commented:
@orangedeng In Linux, the host ports functionality is taken care by "network-manager". Are we not going to run that in Windows environment?

orangedeng commented:
@leodotcloud No, because we can not do `host` network in windows. Also the contains can not modify host network settings. So the only way to make it work is running the binary in windows.

leodotcloud commented:
@yasker @orangedeng For now I prefer to keep the linux releasesseparate from windows. I have created a new branch called "windows", let's merge this code in that branch and make releases from that branch directly. Once things are working end to end we can reevaluate and merge into master. @orangedeng can you please edit the PR to point to windows branch instead of master?

deniseschannon commented:
@leodotcloud No, let's put the releases together. All micro-services are going to have both linux and windows combined.

Commit:
Adapter to AWS EC2

1. Delete route rule for 169.254.169.250 because aws will add this
route when init instance.
2. Find the vitrual NIC with specific name instead of using the
biggest interfaceIndex one.

Commit:
Setup the default dns servers for the container network.

Get nameservers from each validate net adapters and configure them
to per-host-subnet.

leodotcloud commented on routeupdate/hostgw/hostgw_windows.go:
```diff
@@ -0,0 +1,98 @@
+package hostgw
```
Needs: `//+build windows`?

orangedeng replied:
not really needed because of the `_windows` subfix.

leodotcloud commented on routeupdate/hostgw/utils_windows.go:
```diff
@@ -0,0 +1,137 @@
+package hostgw
```
Needs: `//+build windows`?

leodotcloud commented on vendor.conf:
```diff
@@ -7,3 +7,9 @@ github.com/pkg/errors   v0.8.0-1-g839d9e9
+github.com/orangedeng/win-route-netsh f06f27f1fa5955d41f4d344e89283b008900a2cf
+github.com/mitchellh/mapstructure 06020f85339e21b2478f756a78e295255ffa4d6a
+github.com/juju/errors master
+github.com/orangedeng/go-winnat 699b70068ac94a0794a23951f1037fbac947d0af
```
Can we move the dependencies under rancher? instead of orangedeng?

deniseschannon replied:
@orangedeng we should move it to rancher. if you transfer it to me directly, i'll transfer it to rancher

leodotcloud reviewed (changes requested):
Please remove: routeupdate/hostgw/.DS_Store

deniseschannon commented on scripts/package:
```diff
@@ -9,13 +9,29 @@ SUFFIX=""
+
+}
+
+if [ "$CROSS" == "true" ]; then
```
@orangedeng can you update this to `-n "$CROSS" ` like it was updated in rancher/agent so it will produce the correct `.zip` files?

Commit:
Transfer the dependences from orangedeng to rancher

orangedeng commented:
Updated

leodotcloud reviewed (approved):
LGTM
